File: Managed/NextTurn.UE.Runtime/CoreUObject/MapProperty.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public sealed class MapProperty : Property, IProperty<ScriptMap>
    {
        internal MapProperty(IntPtr pointer) : base(pointer) { }

        public Property KeyProperty => Create<Property>(NativeMethods.GetKeyProperty(this.pointer));

        public Property ValueProperty => Create<Property>(NativeMethods.GetValueProperty(this.pointer));

        public ScriptMap GetValue(Object @object, int index = 0) => this.GetValue<ScriptMap>(@object, index);

        public void SetValue(Object @object, ScriptMap value, int index = 0) => this.SetValue<ScriptMap>(@object, value, index);

        internal static new class NativeMethods
        {
            [ReadOffset]
            public static extern IntPtr GetKeyProperty(IntPtr property);

            [ReadOffset]
            public static extern IntPtr GetValueProperty(IntPtr property);

            [ReadOffset]
            public static extern ScriptMap.Layout GetLayout(IntPtr property);
        }
    }
}

File: Source/NextTurnRuntime/Private/CoreUObject/MapProperty.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Accessor.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/UnrealType.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME MapProperty

TYPE_EXPORT_START

const FMemberSymbol Members[] =
{
	EXPORT_OFFSET_PUBLIC(FMapProperty, KeyProperty, KeyProp),
	EXPORT_OFFSET_PUBLIC(FMapProperty, ValueProperty, ValueProp),
	EXPORT_OFFSET_PUBLIC(FMapProperty, Layout, MapLayout),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
